### PR TITLE
Partner Autocompletion

### PR DESCRIPTION
**Pull request by rrahir**
task https://www.odoo.com/web?debug#id=1872667&action=333&active_id=1217&model=project.task&view_type=form&menu_id=4720

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/partner_autocomplete/static/tests/partner_autocomplete_tests.js [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+                formatted = type==='integer' ? parseInt(formatted, 10) : formatted;
+                assert.strictEqual(
+                    formatted,
+                    val === false ? 0 : val,
+                    key + ' value should have been updated to \"' + val + '"'
+                );
+            }
+        });
+
+        $input.val("ZZZZZZZZZZZZZZZZZZZZZZ").trigger("input");
```
Comment (rrahir) [inline addons/partner_autocomplete/static/tests/partner_autocomplete_tests.js]: @twa-odoo bete question: si on a une entrée ZZZZZZZZZZZZZZZZZZZZZZZZ, qu'est-ce qui va se passer? :D on est d'accord que dans la situtation actuelle c'est safe, mais on sait jamais. on devrait ptet avoir un pré process coté client qui vire es suggestions débiles genre "**qé'"'é"é"'é' et qui refuse de faire le rpc dans ce cas-là. je sais pas ce que tu en penses ?

File: addons/partner_autocomplete/static/src/js/partner_autocomplete_many2one.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                return Autocomplete.validateSearchTerm(term)
+            });
+        },
+
+        /**
+         * Query Autocomplete and add results to the popup
+         *
+         * @override
+         * @param search_val
+         * @returns {*}
```
Comment (aab-odoo) [inline addons/partner_autocomplete/static/src/js/partner_autocomplete_many2one.js]: Variable API is bad practice, what about making this function always return a Deferred resolved with an Array, which is sometimes empty?

File: addons/partner_autocomplete/static/tests/partner_autocomplete_tests.js [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    var createView = testUtils.createView;
+
+    QUnit.module('partner_autocomplete', {
+        before: function () {
+            this.__field_clearbit_debounce = FieldAutocomplete.prototype.debounceSuggestions;
+            FieldAutocomplete.prototype.debounceSuggestions = 0;
+
+            // TODO mock these instead of overriding them
+
+            this.__field_clearbit_getBase64Image = FieldAutocomplete.prototype._getBase64Image;
```
Comment (aab-odoo) [inline addons/partner_autocomplete/static/tests/partner_autocomplete_tests.js]: you should use testUtils.patch/unpatch for all you overrides

File: addons/partner_autocomplete/static/tests/partner_autocomplete_tests.js [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                this.suggestions = _.filter(suggestions, function (suggestion) {
+                    return (suggestion.name.toLowerCase().indexOf(value.toLowerCase()) >= 0);
+                });
+                return $.when();
+            };
+
+            this.__field_clearbit_enrichCompany = FieldAutocomplete.prototype._enrichCompany;
+            FieldAutocomplete.prototype._enrichCompany = function () {
+               return {
+                   "country_id": 20,
```
Comment (aab-odoo) [inline addons/partner_autocomplete/static/tests/partner_autocomplete_tests.js]: no need to quote dict keys in JS

File: addons/partner_autocomplete/static/src/js/partner_autocomplete_core.js [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+         */
+        autocomplete: function (value) {
+            if (this._isVAT(value)) {
+                return this.getVatSuggestions(value);
+            } else {
+                return this.getNameSuggestions(value);
+            }
+        },
+
+        getVatSuggestions: function (value) {
```
Comment (aab-odoo) [inline addons/partner_autocomplete/static/src/js/partner_autocomplete_core.js]: don't forget the doctrings

File: addons/web/static/src/js/fields/res_partner_many2one.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+odoo.define('web.res_partner_many2one', function (require) {
+    "use strict";
+
+    var fieldRegistry = require('web.field_registry');
+    var FieldMany2One = require('web.relational_fields').FieldMany2One;
+
+    var PartnerFieldMany2One = FieldMany2One.extend({});
```
Comment (aab-odoo) [inline addons/web/static/src/js/fields/res_partner_many2one.js]: Two questions:
 - for now, there is only one override of this. Do you really need this generic mechanism? Can't you simply create that widget in partner_autocomplete?
 - if the answer of previous question is yes, can't you make FieldMany2one allowing extra autocompletes directly in this widget? Is there a need to make FieldMany2one that generic?

**Pull request closed without merging**